capsules/src/kv_driver.rs

```diff
@@ -0,0 +1,453 @@
+                                                value[..static_buffer_len]
+                                                    .copy_to_slice(&mut buf[..static_buffer_len]);
+
+                                                Ok(())
+                                            })
+                                        })
+                                    })
+                                    .unwrap_or(Err(ErrorCode::RESERVE))?;
+
+                                if let Err((data, dest, e)) = self.kv.set(
+                                    self.data_buffer.take().unwrap(),
+                                    self.dest_buffer.take().unwrap(),
+                                ) {
+                                    self.data_buffer.replace(data);
+                                    self.dest_buffer.replace(dest);
+                                    return e;
+                                }
+                            }
+                            UserSpaceOp::Delete => {
+                                kernel_data
+                                    .get_readonly_processbuffer(ro_allow::UNHASHED_KEY)
+                                    .and_then(|buffer| {
+                                        buffer.enter(|unhashed_key| {
+                                            let mut static_buffer_len = 0;
+                                            self.data_buffer.map_or(Err(ErrorCode::NOMEM), |buf| {
+                                                // Determine the size of the static buffer we have
+                                                static_buffer_len = buf.len();
+
+                                                if static_buffer_len > unhashed_key.len() {
+                                                    static_buffer_len = unhashed_key.len()
+                                                }
+
+                                                // Copy the data into the static buffer
+                                                unhashed_key[..static_buffer_len]
+                                                    .copy_to_slice(&mut buf[..static_buffer_len]);
+
+                                                Ok(())
+                                            })
+                                        })
+                                    })
+                                    .unwrap_or(Err(ErrorCode::RESERVE))?;
+
+                                if let Err((data, e)) =
+                                    self.kv.delete(self.data_buffer.take().unwrap())
+                                {
+                                    self.data_buffer.replace(data);
+                                    return e;
+                                }
+                            }
+                        }
+                    }
+
+                    Ok(())
+                })
+                .unwrap_or_else(|err| Err(err.into()))
+        })
+    }
+
+    fn check_queue(&self) {
```

Comment:
This seems out of place here, but, I feel like you've come this far, maybe it's worth doing the userspace virtualization too?